File: algorithms/0005.longest-palindromic-substring/solution.rs
impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        let chars: Vec<char> = s.chars().collect();
        let (mut left, mut right) = (0, 0);
        let mut longest = 0;
        for i in 0..(s.len() - 1) {
            for j in (i + 1)..s.len() {
                if longest > j - i {
                    continue;
                }

                if Self::is_palindrome(&chars, i, j) {
                    longest = j - i;
                    left = i;
                    right = j;
                }
            }
        }

        s[left..=right].to_string()
    }

    fn is_palindrome(chars: &Vec<char>, mut left: usize, mut right: usize) -> bool {
        while left < right {
            if chars[left] != chars[right] {
                return false;
            }

            left += 1;
            right -= 1;
        }

        true
    }
}

File: algorithms/0005.longest-palindromic-substring/solution.cpp
class Solution {
    bool isPalindrome(string& s, int left, int right) {
        while (left < right) {
            if (s[left] != s[right]) {
                return false;
            }

            left++;
            right--;
        }
        return true;
    }

  public:
    string longestPalindrome(string s) {
        int len = s.length();
        int longest = 0;
        int left = 0;
        int right = 0;

        for (int i = 0; i < len - 1; i++) {
            for (int j = i + 1; j < len; j++) {
                if (longest > j - i) {
                    continue;
                }

                if (isPalindrome(s, i, j)) {
                    longest = j - i;
                    left = i;
                    right = j;
                }
            }
        }

        return s.substr(left, right - left + 1);
    }
};
